internal/controller/cull.go

```diff
@@ -41,36 +44,170 @@ func updateHibernationState(ctx context.Context, r *AmaltheaSessionReconciler, a
+	}
+	return metav1.Time{}
+}
+
+func hibernationDateByStartingDuration(creationTimestamp metav1.Time, status amaltheadevv1alpha1.AmaltheaSessionStatus, culling amaltheadevv1alpha1.Culling) metav1.Time {
+	zero := time.Duration(0)
+	maxStartingDuration := culling.MaxStartingDuration.Duration
+	starting := status.State == amaltheadevv1alpha1.NotReady
+	if maxStartingDuration > zero && starting {
+		return metav1.NewTime(creationTimestamp.Add(maxStartingDuration))
+	}
+	return metav1.Time{}
+}
+
+func hibernationDateByFailingSince(status amaltheadevv1alpha1.AmaltheaSessionStatus, culling amaltheadevv1alpha1.Culling) metav1.Time {
+	zero := time.Duration(0)
+	failingSince := status.FailingSince
+	maxFailedDuration := culling.MaxFailedDuration.Duration
+	if !failingSince.IsZero() && maxFailedDuration > zero {
+		return metav1.NewTime(failingSince.Add(maxFailedDuration))
+	}
+	return metav1.Time{}
+}
+
+func hibernationDateByIdleSince(status amaltheadevv1alpha1.AmaltheaSessionStatus, culling amaltheadevv1alpha1.Culling) metav1.Time {
+	zero := time.Duration(0)
+	idleSince := status.IdleSince
+	lastIdleSince := idleSince
+	maxIdleDuration := culling.MaxIdleDuration.Duration
+	// if we have a last-interaction, it determines the start of the idle
+	// time that is used to decide upon hibernation
+	lastInteraction := culling.LastInteraction
+	if !lastInteraction.IsZero() && !idleSince.IsZero() && lastInteraction.After(idleSince.Time) {
+		lastIdleSince = lastInteraction
+	}
+	if !lastIdleSince.IsZero() && maxIdleDuration > zero {
+		return metav1.NewTime(lastIdleSince.Time.Add(maxIdleDuration))
+	}
+	return metav1.Time{}
+}
+
+func calculateHibernationDate(log logr.Logger, creationTimestamp metav1.Time, status amaltheadevv1alpha1.AmaltheaSessionStatus, culling amaltheadevv1alpha1.Culling) metav1.Time {
+	type Pair struct {
+		Name string
+		Date metav1.Time
+	}
+
+	hibernationDates := []Pair{
+		{Name: "MaxAge", Date: hibernationDateByMaxAge(creationTimestamp, culling)},
+		{Name: "StartDuration", Date: hibernationDateByStartingDuration(creationTimestamp, status, culling)},
+		{Name: "FailingDuration", Date: hibernationDateByFailingSince(status, culling)},
+		{Name: "IdleDuration", Date: hibernationDateByIdleSince(status, culling)},
+	}
+
+	result := Pair{}
+	logMsg := ""
+	for _, hd := range hibernationDates {
+		if result.Date.IsZero() || (!hd.Date.IsZero() && hd.Date.Time.Before(result.Date.Time)) {
+			result = hd
+		}
+		if !hd.Date.IsZero() {
+			logMsg += fmt.Sprint(" ", hd.Name, ":", hd.Date)
+		}
+	}
+	if result.Date.IsZero() {
+		logMsg += " => NONE"
+	} else {
+		logMsg += fmt.Sprint(" => ", result.Name, ":", result.Date)
+	}
+	log.Info("hibernationDate", "DecisionLog", strings.TrimSpace(logMsg))
+	return result.Date
+}
+
+func needsScaleDown(hibernationDate metav1.Time) bool {
+	now := time.Now()
+	return !hibernationDate.IsZero() && (hibernationDate.Time.Equal(now) || hibernationDate.Time.Before(now))
```

Comment:
The case where equality would happen is vanishingly small, so we do not need to have it:

```suggestion
	return !hibernationDate.IsZero() && hibernationDate.Time.Before(now)
```